### PR TITLE
Switch Swap Provider

### PR DESCRIPTION
**Pull request by FrederikBolding**

### CODE REVIEW

**Comment by FrederikBolding** [comment]
[ch7904]

**Comment by FrederikBolding** [comment]
Waiting on a small styling fix from #3823 but otherwise good to review.

File: src/services/ApiService/Dex/Dex.ts [hunk 1]
```diff
@@ -42,154 +53,103 @@ export default class DexService {
   }
 
   public getTokenList = async (): Promise<DexAsset[]> => {
-    const { data: tokenList }: { data: DexAsset[] } = await this.service.get('token-list-full');
+    const {
+      data: { records: tokenList }
+    }: { data: { records: DexAsset[] } } = await this.service.get('swap/v1/tokens');
```
Comment (vilsbole) [inline src/services/ApiService/Dex/Dex.ts:58]: What happens when the call fails and `data` is undefined?
A try/catch or return the Promise ?

File: src/hooks/useTxMulti/useTxMulti.ts [hunk 2]
```diff
@@ -27,7 +27,7 @@ export type TUseTxMulti = () => {
   state: TxMultiState;
   init(txs: any[], account: any, network: any): Promise<void>;
   initWith(
-    getTxs: () => Promise<Partial<ITxObject & { label: string; type: ITxType }>[]>,
+    getTxs: () => Promise<Partial<ITxObject & { label?: string; type?: ITxType }>[]>,
```
Comment (vilsbole) [inline src/hooks/useTxMulti/useTxMulti.ts:30]: Duplicate ts indicator Partial already makes the keys optional

File: src/helpers/erc20.ts [hunk 3]
```diff
@@ -14,13 +14,12 @@ import { inputValueToHex } from '@utils';
 interface IFormatApproveTxInputs {
   contractAddress: ITxToAddress;
   baseTokenAmount: Bigish;
-  fromAddress: ITxFromAddress;
+  fromAddress?: ITxFromAddress;
```
Comment (vilsbole) [inline src/helpers/erc20.ts:17]: When is this optional?

File: src/features/SwapAssets/components/SwapAssets.tsx [hunk 4]
```diff
@@ -147,17 +161,56 @@ export const SwapAssets = (props: Props) => {
...
+          <Body>
+            {translateRaw('ACCOUNT_SELECTION_PLACEHOLDER')}{' '}
+            <Tooltip tooltip={translateRaw('SWAP_SELECT_ACCOUNT_TOOLTIP')} />
+          </Body>
+        </Box>
+        <AccountSelector
+          name="account"
+          value={account}
+          accounts={filteredAccounts}
+          onSelect={(option: StoreAccount) => {
```
Comment (vilsbole) [inline src/features/SwapAssets/components/SwapAssets.tsx:201]: Simplify?
`onSelect={handleAccountSelected}`

**Review by vilsbole — CHANGES_REQUESTED** [review]

**Comment by vilsbole** [comment]
Place refresh in line with the expire ? 

Screenshot 2021-02-11 at 17 21 34

File: src/helpers/erc20.ts (hunk 3 above)
Comment (FrederikBolding) [reply src/helpers/erc20.ts:17]: When we are creating an approval tx for swap but don't know the sender yet fx.

File: src/services/ApiService/Dex/Dex.ts (hunk 1 above)
Comment (FrederikBolding) [reply src/services/ApiService/Dex/Dex.ts:58]: They are wrapped in try catch when used, inside the stateFactory. Do you want to do something other than that?

**Comment by FrederikBolding** [comment]
> Place refresh in line with the expire ?
> 
> Screenshot 2021-02-11 at 17 21 34

I don't understand what you are asking for here

File: src/helpers/erc20.ts (hunk 3 above)
Comment (vilsbole) [reply src/helpers/erc20.ts:17]: very well

File: src/services/ApiService/Dex/Dex.ts (hunk 1 above)
Comment (vilsbole) [reply src/services/ApiService/Dex/Dex.ts:58]: Fine by me

File: src/components/Amount.tsx [hunk 5]
```diff
@@ -52,6 +52,7 @@ interface Props {
   baseAssetValue?: string;
   fiatColor?: string;
   bold?: boolean;
+  style?: CSSProperties;
```
Comment (vilsbole) [inline src/components/Amount.tsx:55]: We should avoid redeclaring react component props.
`const Amount: React.FC<Props> = ({}) => {}` is sufficient to type `style`

File: src/config/data.tsx [hunk 6]
```diff
@@ -35,12 +35,9 @@ export const SUPPORT_EMAIL = 'support@mycrypto.com';
...
 export const CRYPTOSCAMDB = 'https://cryptoscamdb.org';
 
-// Handler address will change if the trade contract changes.
-// Can also access handler by calling `approvalHandler()` on the trade contract.
-// Also update eth contracts when updating DexAG addresses
-export const DEXAG_MYC_TRADE_CONTRACT = '0x3d7b19C37d422B43c07C7Ba6353ED2D1689540FD'; //'0xA65440C4CC83D70b44cF244a0da5373acA16a9cb';
-export const DEXAG_MYC_HANDLER_CONTRACT = '0xae0A30ac8C76d85f1E76A3EbDcEdc3047e0da456';
-export const DEX_BASE_URL = 'https://api-v2.dex.ag/';
+export const DEX_BASE_URL = 'https://api.0x.org/';
+export const DEX_FEE_RECIPIENT = '0x4bbeEB066eD09B7AEd07bF39EEe0460DFa261520'; // @todo Set to an address we want
```
Comment (vilsbole) [inline src/config/data.tsx:39]: @tayvano Which address should we use?

File: src/config/data.tsx [hunk 7]
```diff
@@ -35,12 +35,9 @@ export const SUPPORT_EMAIL = 'support@mycrypto.com';
...
+export const DEX_TRADE_EXPIRATION = 31;
```
Comment (vilsbole) [inline src/config/data.tsx:40]: Comment to specify the value units

File: src/features/SwapAssets/SwapAssetsFlow.tsx [hunk 8]
```diff
@@ -142,7 +158,7 @@ const SwapAssetsFlow = (props: RouteComponentProps) => {
         }
       },
       {
-        title: translateRaw('SWAP'),
+        title: '',
```
Comment (vilsbole) [inline src/features/SwapAssets/SwapAssetsFlow.tsx:161]: Link to review comment to specify duplicate title issue.

File: src/features/SwapAssets/components/SwapAssets.tsx [hunk 9]
```diff
@@ -147,17 +156,65 @@ export const SwapAssets = (props: Props) => {
...
+
+  const estimatedGasFee =
+    gasPrice &&
+    tradeGasLimit &&
+    totalTxFeeToString(
+      gasPrice,
+      bigify(tradeGasLimit).plus(approvalGasLimit ? approvalGasLimit : 0)
+    );
+
+  const [isExpired, setIsExpired] = useState(false);
```
Comment (vilsbole) [inline src/features/SwapAssets/components/SwapAssets.tsx:171]: Move state declaration to top of component.

File: src/features/SwapAssets/components/SwapAssets.tsx [hunk 10]
```diff
@@ -83,22 +81,33 @@ export const SwapAssets = (props: Props) => {
...
+    isEstimatingGas,
+    expiration,
     isDemoMode
   } = props;
 
   const { accounts, userAssets } = useContext(StoreContext);
+  const { getNetworkById } = useNetworks();
+  const { assets: allAssets } = useAssets();
+  const { getAssetRate } = useRates();
+  const { settings } = useSettings();
```
Comment (vilsbole) [inline src/features/SwapAssets/components/SwapAssets.tsx:100]: Replace hooks with `mapStateToProps`.
There is no reason to not do it now.

**Review by vilsbole — CHANGES_REQUESTED** [review]

File: src/components/Amount.tsx (hunk 5 above)
Comment (FrederikBolding) [reply src/components/Amount.tsx:55]: That doesn't seem to be the case 😆

Comment (vilsbole) [reply src/components/Amount.tsx:55]: My bad! `FC` doesn't add style attribute. Closing.

**Review by vilsbole — COMMENTED** [review]
code lgtm;

**Review by vilsbole — DISMISSED** [review]

**Comment by blurpesec** [comment]
I set up the mycryptodev version of this and was testing it. Can we capture the `code: 109` error response and display "Insufficient ETH for transaction fees" or something like that instead of just displaying the error message they send back? The actual result sent back was: `{"code":109,"reason":"Insufficient funds for transaction"}`

image

File: src/features/SwapAssets/stateFormFactory.tsx [hunk 11]
```diff
@@ -40,6 +34,7 @@ const swapFormInitialState = {
 
 const SwapFormFactory: TUseStateReducerFactory<SwapFormState> = ({ state, setState }) => {
   const { networks } = useNetworks();
+  const network = getNetworkById(DEFAULT_NETWORK, networks);
```
Comment (vilsbole) [inline src/features/SwapAssets/stateFormFactory.tsx:37]: Use store selector.
`useSelector(getNetwork(DEFAULT_NETWORK))`

**Review by vilsbole — CHANGES_REQUESTED** [review]

File: src/features/SwapAssets/components/SwapAssets.tsx [hunk 12]
```diff
@@ -83,46 +73,63 @@ export const SwapAssets = (props: Props) => {
...
 
+  const [isExpired, setIsExpired] = useState(false);
   const { accounts, userAssets } = useContext(StoreContext);
+  const { getAssetRate } = useRates();
 
-  // Accounts with a balance of the chosen asset
-  const filteredAccounts = fromAsset
-    ? getAccountsWithAssetBalance(accounts, fromAsset, fromAmount)
-    : [];
+  const baseAsset = getBaseAssetByNetwork({ network, assets: allAssets })!;
```
Comment (vilsbole) [inline src/features/SwapAssets/components/SwapAssets.tsx:95]: Define selector and pass with connect.

**Review by vilsbole — CHANGES_REQUESTED** [review]

**Review by vilsbole — APPROVED** [review]
code lgtm;

**Pull request merged**